File: e2e/java/src/test/java/dev/kreuzberg/e2e/EncodingTest.java
package dev.kreuzberg.e2e;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import dev.kreuzberg.kreuzcrawl.Kreuzcrawl;

/** E2e tests for category: encoding. */
class EncodingTest {
    @Test
    void testEncodingDoubleEncoded() throws Exception {
        // Handles double-encoded URL characters (%25C3%25B6)
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertFalse(result.html().isEmpty(), "expected non-empty value");
        assertTrue(result.links().size() >= 1, "expected >= 1");
    }

    @Test
    void testEncodingMixedCharsetPage() throws Exception {
        // Handles charset mismatch between HTTP header and HTML meta tag
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertFalse(result.html().isEmpty(), "expected non-empty value");
    }

    @Test
    void testEncodingPercentEncodedPath() throws Exception {
        // Handles percent-encoded spaces and characters in URL paths
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertFalse(result.html().isEmpty(), "expected non-empty value");
        assertTrue(result.links().size() >= 2, "expected >= 2");
    }

    @Test
    void testEncodingUnicodeUrl() throws Exception {
        // Handles Unicode characters in URLs (Hebrew, Japanese, Cyrillic)
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertFalse(result.html().isEmpty(), "expected non-empty value");
    }

}

File: e2e/rust/tests/encoding_test.rs
//! E2e tests for category: encoding

use kreuzcrawl::scrape;
use kreuzcrawl::create_engine;

#[tokio::test]
async fn test_encoding_double_encoded() {
    // Handles double-encoded URL characters (%25C3%25B6)
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert!(!result.html.is_empty(), "expected non-empty value");
    assert!(result.links.len() >= 1, "expected >= 1");
}

#[tokio::test]
async fn test_encoding_mixed_charset_page() {
    // Handles charset mismatch between HTTP header and HTML meta tag
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert!(!result.html.is_empty(), "expected non-empty value");
}

#[tokio::test]
async fn test_encoding_percent_encoded_path() {
    // Handles percent-encoded spaces and characters in URL paths
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert!(!result.html.is_empty(), "expected non-empty value");
    assert!(result.links.len() >= 2, "expected >= 2");
}

#[tokio::test]
async fn test_encoding_unicode_url() {
    // Handles Unicode characters in URLs (Hebrew, Japanese, Cyrillic)
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert!(!result.html.is_empty(), "expected non-empty value");
}
